### ISSUE TITLE
In platform Threads there "is no context switch"?

### ISSUE THREAD

**Issue by fjf2002** [jesperancinha/good-story#87]
Hello Joao,

I've read your article at https://itnext.io/kotlin-coroutines-vs-java-virtual-threads-a-good-story-but-just-that-91038c7d21eb .
I couldn't find out how to connect with you otherwise, so I'm trying it here.

In your article, you say,

> In a concurrent environment with just the use of just of platform Threads, **there is no context switch**, and therefore making a blocking call **always means waiting for the blocking call to finish** before being allowed to continue. Coroutines or Continuations explore threads to the maximum by making sure that we avoid anything to block whenever that is possible.

Well, wrong. Or at least misleading. The thing is, in a multitasking operating system, we have
* processes,
* threads,
* and now in both kotlin and java: coroutines (aka virtual threads aka fibers).

**All** of these execution models are being managed by some scheduler, and on blocking operations, control (i. e. the CPU) is given away to others.

The difference is, processes and threads are managed by the OS, whereas the coroutines discussed are managed by the Java/Kotlin compiler and/or runtime.

From a programmer's standpoint, one can partly abstract away **who** manages his multitasking. So at first glance, "inventing" coroutines in Java/Kotlin seems like re-inventing the wheel. The sole point is, in my judgement: coroutines are more lightweight than threads, just as threads are more lightweight than processes.

P. S.: There already happened a step in previous Java versions to make multithreading more lightweight: The introduction of thread pools, and the idea that so-called executors distribute runnables across the thread pool. This lowered thread construction/termination costs (as opposed to constructing one thread for each runnable).

P. P. S.: It is interesting to see that browsers kind of went the other way: browser tabs are running in separate processes, for better isolation (security, and perhaps stability).

**Comment by jesperancinha** [comment]
Hi there @fjf2002,

First of all thank your for reading the article and for giving your opinion. I don't know why, but in IT it's not that comment to see people making comments and posts about articles. This is why I think it's great that you've opened this issue for me.

I wrote this article without any help, incentives or whatever and my focus was to understand really quick how coroutines work. I wasn't even aware that Virtual Threads are part of the JVM and that kotlin coroutines were actually implemented in the standard library. In broad terms that is... Anyways, I do mention in the article that it is a very experimental one and the specific error you mention, I completely acknowledge that. Threads can switch context as well of course.

I love that fact that you mention that `coroutines` are more lightweight than `threads` and `threads` are more lightweight than `processes`. The whole article is focused on low level code and really going into the entrais of the implementation, but I think I still only scratched the surface with it and a lot is left to explore.

An example of this is that I'm comparing the Dispatcher.Default launched coroutines with virtual threads. This doesn't make sense and that is actually the reason why in my readings I got the number of cores for coroutines and a huge number for simultaneous virtual threads by comparison.

So there are a few points I definitely need to refactor in that article:

* kotlin coroutines vs threads vs processes
* continuations
* correct the comparison and re-run the tests.

Thanks you for your comment and contribution. I've set this issue on my name and I will re-edit the article soon!
